File: src/main/java/org/radarbase/management/security/SecurityUtils.java
package org.radarbase.management.security;

import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContext;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;

import java.util.Optional;

/**
 * Utility class for Spring Security.
 */
public final class SecurityUtils {
    private SecurityUtils() {
    }

    /**
     * Get the login of the current user.
     *
     * @return the login of the current user if present
     */
    public static Optional<String> getCurrentUserLogin() {
        SecurityContext securityContext = SecurityContextHolder.getContext();
        return getUserName(securityContext.getAuthentication());
    }

    /**
     * Get the user name contianed in an Authentication object.
     *
     * @param authentication context authentication
     * @return user name if present
     */
    public static Optional<String> getUserName(Authentication authentication) {
        return Optional.ofNullable(authentication)
                .map(Authentication::getPrincipal)
                .map(principal -> {
                    if (principal instanceof UserDetails) {
                        return ((UserDetails) authentication.getPrincipal()).getUsername();
                    } else if (principal instanceof String) {
                        return (String) authentication.getPrincipal();
                    } else if (principal instanceof Authentication) {
                        return ((Authentication)principal).getName();
                    } else {
                        return null;
                    }
                });
    }
}

File: src/main/java/org/radarbase/management/security/SecurityUtils.kt
package org.radarbase.management.security

import org.springframework.security.core.Authentication
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.core.userdetails.UserDetails

/**
 * Utility class for Spring Security.
 */
object SecurityUtils {
    val currentUserLogin: String?
        /**
         * Get the login of the current user.
         *
         * @return the login of the current user if present
         */
        get() {
            val securityContext = SecurityContextHolder.getContext()
            return getUserName(securityContext.authentication)
        }

    /**
     * Get the user name contianed in an Authentication object.
     *
     * @param authentication context authentication
     * @return user name if present
     */
    fun getUserName(authentication: Authentication): String? = authentication
        .let { obj: Authentication -> obj.principal }
        .let { principal: Any? ->
            when (principal) {
                is UserDetails -> {
                    return (authentication.principal as UserDetails).username
                }

                is String -> {
                    return authentication.principal as String
                }

                is Authentication -> {
                    return principal.name
                }

                else -> {
                    return null
                }
            }
        }
}
